File: native/android/app/src/main/java/app/stapes/Haptics.kt
package app.stapes

import android.content.Context
import android.os.Build
import android.os.VibrationEffect
import android.os.Vibrator
import android.os.VibratorManager
import android.webkit.JavascriptInterface

/**
 * The feel of a blow, and the bridge the page reaches it through.
 *
 * **The web side names the event and this decides the waveform.** @see
 * `app/lib/haptics.ts`, and `native/ios/Stapes/Haptics.swift` for the same
 * decision made with the other platform's vocabulary.
 *
 * Installed as `window.StapesNative`. Everything annotated `@JavascriptInterface`
 * here is callable by any script the page runs, so this object stays exactly as
 * wide as a vibration and no wider — which is also why `MainActivity` refuses to
 * navigate off the origin: the bridge is only as trustworthy as the page it is
 * handed to.
 */
class Haptics(context: Context) {
    private val vibrator: Vibrator? = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
        val manager = context.getSystemService(VibratorManager::class.java)
        manager?.defaultVibrator
    } else {
        @Suppress("DEPRECATION")
        context.getSystemService(Vibrator::class.java)
    }

    /**
     * Play one.
     *
     * Called on a background thread — that is where the WebView delivers every
     * `@JavascriptInterface` call — which the vibrator is happy with and which
     * is why nothing here touches a view.
     *
     * `intensity` is how much of the body the blow took, 0 to 1, and is floored
     * rather than passed straight through: a buzz too faint to feel reads as
     * broken haptics rather than as a scratch. Devices without amplitude
     * control ignore the number and give their one strength, which is the right
     * degradation.
     */
    @JavascriptInterface
    fun haptic(kind: String, intensity: Double) {
        if (kind != "hit") return
        val device = vibrator ?: return
        if (!device.hasVibrator()) return
        val strength = intensity.coerceIn(0.35, 1.0)
        val amplitude = (strength * 255).toInt().coerceIn(1, 255)
        device.vibrate(VibrationEffect.createOneShot(BLOW_MS, amplitude))
    }

    private companion object {
        /**
         * Short enough to read as an impact rather than as a notification. The
         * platform has no stock "something hit you" effect — `EFFECT_CLICK` is a
         * tick for a control, and the predefined set is about interface
         * feedback rather than about events in a world — so this is the one
         * number the Android side has to choose.
         */
        const val BLOW_MS = 28L
    }
}

File: app/lib/haptics.ts
/**
 * A vibration, played by the native shell when the game is running inside one.
 *
 * **Nothing here reaches the browser's own `navigator.vibrate`, on purpose.**
 * Safari ignores it, Chrome on Android does not, and the result would be that
 * the same game buzzes in one mobile browser and not the other — a difference
 * nobody chose and nobody can turn off. A shell that offers haptics has opted
 * in by installing a bridge; a tab has not.
 *
 * The web side names *what happened* and never the waveform. Both platforms
 * have a vocabulary of stock feedback patterns tuned to their own hardware —
 * `UIImpactFeedbackGenerator` on iOS, `VibrationEffect` on Android — and a
 * duration in milliseconds picked here would throw all of that away to be worse
 * on both. See `native/ios/Stapes/Haptics.swift` and
 * `native/android/app/src/main/java/app/stapes/Haptics.kt`, which is where the
 * feel of each of these is decided.
 */

/** The things worth feeling. One, so far. */
export type HapticKind = "hit";

/**
 * The shells, as the page sees them.
 *
 * Two mechanisms because the platforms offer two: WebKit gives a page a message
 * handler per name registered by the host, and Android injects a named object
 * with methods on it. Neither is polyfillable into the other and both are three
 * lines, so the bridge speaks whichever it finds.
 */
declare global {
  interface Window {
    webkit?: {
      messageHandlers?: {
        stapesHaptics?: { postMessage(body: unknown): void };
      };
    };
    StapesNative?: { haptic(kind: string, intensity: number): void };
  }
}

/**
 * Is a native shell listening?
 *
 * Exported because the answer is worth having before doing work for a
 * vibration that nobody will feel, and because the two bridges are an
 * implementation detail that should stay in this file.
 */
export function hasHaptics(): boolean {
  if (typeof window === "undefined") return false;
  return Boolean(
    window.webkit?.messageHandlers?.stapesHaptics || window.StapesNative,
  );
}

/**
 * Play one, if there is anything to play it on.
 *
 * `intensity` is 0 to 1 and is a hint rather than an instruction: it is how
 * much of *this body* the blow took, so a scratch and a near-fatal hit are not
 * the same thump. What each platform does with it — scale an impact, pick a
 * different stock effect, or ignore it — belongs to the shell.
 *
 * Never throws. A vibration is the least important thing happening on the frame
 * it is asked for, and a bridge that has gone away mid-session must not take
 * the tick handler down with it.
 */
export function haptic(kind: HapticKind, intensity = 1): void {
  if (typeof window === "undefined") return;
  // `Math.min` and `Math.max` both pass `NaN` straight through, and a division
  // by a `maxHp` of zero is one arithmetic slip away at the only call site. The
  // shells would hand it to `impactOccurred(intensity:)` and to an amplitude
  // between 1 and 255, neither of which has an answer for it.
  const clamped = Number.isFinite(intensity)
    ? Math.min(1, Math.max(0, intensity))
    : 1;
  try {
    const ios = window.webkit?.messageHandlers?.stapesHaptics;
    if (ios) {
      ios.postMessage({ kind, intensity: clamped });
      return;
    }
    window.StapesNative?.haptic(kind, clamped);
  } catch {
    // A shell that removed its handler, or one whose bridge object is a
    // stranger's. Silence is the right outcome either way.
  }
}
